Explicitly encode output as UTF-8

nickjwhite opened the pull request:
I think I've covered all the cases we need to, but it's possible there could be more.

This should address issue #197 and #10.

kba commented:
These will only change the debug output. Python3 should already do this but won't break with it. LGTM. @zuphilip ?

QuLogic commented:
`str` + `bytes` does not work; this is broken on Python 3.

nickjwhite commented:
Thanks @QuLogic, I don't write a lot of Python, so hadn't spotted the mixing of str + bytes.

I just updated the pull request, with a version that ensures nothing but str is in the print() arguments. And to force UTF-8, I switched to using codecs.getwriter('utf8') for stdout and stderr for all cases where they could output it. I also improved the commit message, so it's clearer what the problem is and why it's worth fixing.

nickjwhite commented:
Just as an additional justification / use-case for this being necessary, I'm currently using Ocropus on a system (that I don't manage) which uses slurm, and no matter what I try I cannot persuade the slurm job manager on this system to not strip locale information. And again, to reiterate, UTF-8 is always the correct thing to output here, regardless of locale, and no fatal errors should be emitted just because of a weird locale.

QuLogic commented:
Does setting ['PYTHONIOENCODING`](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONIOENCODING) not help?

kba commented:
As laid out in #197 there are a few options to achieve this. Revisiting the UTF8/debug output issues, I'm not keen on explicitly adding boilerplate to every file with print statements. Changing stdout/stderr encoding in a single place (like common.py) probably achieves the same goal but is error-prone.

I would prefer to replace print/print_info/print_error with a dedicated logging mechanism, such as python's bundled `logging` or something smaller in the codebase. That would make it easier to override behavior (such as output stream, encoding, logging level).
